content/en/docs/configuration/acme/dns01/acme-dns.md

```diff
@@ -31,85 +30,141 @@ Information about setting up and configuring ACMEDNS is available on the
-  `fulldomain` field in the registration:
+   `fulldomain` field in the registration:
 
-```
-_acme-challenge.example.com CNAME d420c923-bbd7-4056-ab64-c3ca54c9b3cf.auth.example.com
-_acme-challenge.example.org CNAME d420c923-bbd7-4056-ab64-c3ca54c9b3cf.auth.example.com
-```
+    ```
+    _acme-challenge.example.com CNAME d420c923-bbd7-4056-ab64-c3ca54c9b3cf.auth.example.com
+    _acme-challenge.example.org CNAME d420c923-bbd7-4056-ab64-c3ca54c9b3cf.auth.example.com
+    ```
+
+    The "name" of the record always has the _acme-challenge subdomain, and
+    the "value" of the record matches exactly the fulldomain field from
+    registration.
 
-> Note: that the "name" of the record always has the `_acme-challenge`
-> subdomain, and the "value" of the record matches exactly the `fulldomain`
-> field from registration.
+    At verification time, the domain name `d420c923-bbd7-4056-ab64-c3ca54c9b3cf.auth.example.com` will be a TXT
+    record that is set to your validation token. When the verifier queries `_acme-challenge.example.com`, it will
+    be directed to the correct location by this CNAME record. This proves that you control `example.com`
 
-At verification time, the domain name `d420c923-bbd7-4056-ab64-c3ca54c9b3cf.auth.example.com` will be a TXT
-record that is set to your validation token. When the verifier queries `_acme-challenge.example.com`, it will
-be directed to the correct location by this CNAME record. This proves that you control `example.com`
+4. Create a secret from the credentials JSON that was saved in step 2, this
+   secret is referenced in the `accountSecretRef` field of your DNS01
+   issuer settings. When creating an `Issuer` both this `Issuer` and
+   `Secret` must be in the same namespace. However for a `ClusterIssuer`
+   (which does not have a namespace) the `Secret` must be placed in the
+   same namespace as where the cert-manager pod is running in (in the
+   default setup `cert-manager`).
 
-4. Create a secret from the credentials JSON that was saved in step 2, this secret is referenced
-   in the `accountSecretRef` field of your DNS01 issuer settings.
-   When creating an `Issuer` both this `Issuer` and `Secret` must be in the same namespace.
-   However for a `ClusterIssuer` (which does not have a namespace) the `Secret` must be placed in
-   the same namespace as where the cert-manager pod is running in (in the default setup `cert-manager`).
+   ```sh
+   kubectl create secret generic acme-dns --from-file acmedns.json
+   ```
 
-```bash
-$ kubectl create secret generic acme-dns --from-file acmedns.json
+## Limitation of the `acme-dns` server
+
+The [`acme-dns`](https://github.com/joohoi/acme-dns) server has a [known
+limitation](https://github.com/jetstack/cert-manager/issues/3610): the same
+subdomain cannot be used multiple times in the same certificate. For
+example, the following certificate requires two separate challenges that
+will be issued simultaneously using the `acmeDNS` solver:
```

Comment:
From reading the bug it sounds like the issue is when you need three or more TXT records for the same subdomain?

https://github.com/jetstack/cert-manager/issues/3610#issuecomment-782007912

Reply:
You are right, I think I misunderstood the statement in [acme-dns's README](https://github.com/joohoi/acme-dns) and I'm glad someone caught me on that 😅 I went a bit fast and mis-read the OP's comments...

I have been trying to wrap my head around this for 30 minutes and it seems less obvious than what I initially thought... I'll dig a bit more

Reply:
Douple-checking in [db.go](https://github.com/joohoi/acme-dns/blob/68bb6ab654b6fb1fe375e08807688c55621513a2/db.go#L256) (simplified for the purpose of demonstration), acme-dns does in fact return at most two TXT records for one given domain:

```sh
func GetTXTForDomain(domain string) ([]string, error) {
    var txts []string
    q := db.Prepare(`SELECT Value FROM txt WHERE Subdomain=$1 LIMIT 2`) // 🔥
    rows, _ := q.Query(domain)
    for rows.Next() {
        txts = append(txts, rows.Scan())
    }
    return txts, nil
}
```

Imagining that I am running `acme-dns` on 81.82.83.84, and if I want to have certificate for the subdomains `mael.pw`, `blog.mael.pw` and `minio.mael.pw` using a single acme-dns credential (`da81d9b9-4ed1-5343-9a44-4e78f9ead5d9`), my nameserver would look like this:

```sh
NAME                              TYPE    TTL    DATA
acme-dns.mael.pw.                 NS      300    acme-dns.mael.pw.
acme-dns.mael.pw.                 A       300    81.82.83.84
_acme-challenge.mael.pw.          CNAME   300    da81d9b9-4ed1-5343-9a44-4e78f9ead5d9.acme-dns.mael.pw
_acme-challenge.blog.mael.pw.     CNAME   300    da81d9b9-4ed1-5343-9a44-4e78f9ead5d9.acme-dns.mael.pw
_acme-challenge.minio.mael.pw.    CNAME   300    da81d9b9-4ed1-5343-9a44-4e78f9ead5d9.acme-dns.mael.pw
```

I then create this Certificate:

```yaml
kind: Certificate
spec:
  issuerRef: acme-dns-mael-pw
  dnsNames:
    - mael.pw
    - blog.mael.pw
    - minio.mael.pw
```

cert-manager will first do a call to `acme-dns.mael.pw` to add the TXT records. The TXT records in my acme-dns's database will look like:

```sh
NAME                              TYPE  TTL    DATA
da81d9b9-4ed1-5343-9a44-4e78f9ead5d9.acme-dns.mael.pw.      TXT   300    D-XoblaY4uYM-PO8FrRcbRj45cQF-aYluYM4Xob
da81d9b9-4ed1-5343-9a44-4e78f9ead5d9.acme-dns.mael.pw.      TXT   300    D-52Wm4V7xoUpGax-F8FrPO45cQRcbRj-XoblaY4uYM
da81d9b9-4ed1-5343-9a44-4e78f9ead5d9.acme-dns.mael.pw.      TXT   300    D-UpGxoax52Wm4V7-52Wm4V7xoUpGax-Y4uYXobMla
```

Let's Encrypt then proceeds with trying out the challenges. Since ACME servers have to follow CNAME records, the resolution of `_acme-challenge.mael.pw.` should looks like this:

```sh
_acme-challenge.mael.pw.
└── CNAME da81d9b9-4ed1-5343-9a44-4e78f9ead5d9.acme-dns.mael.pw
    ├── TXT D-XoblaY4uYM-PO8FrRcbRj45cQF-aYluYM4Xob
    ├── TXT D-52Wm4V7xoUpGax-F8FrPO45cQRcbRj-XoblaY4uYM
    └── TXT D-UpGxoax52Wm4V7-52Wm4V7xoUpGax-Y4uYXobMla
```

Let's Encrypt (like any other ACME server) is able to work with multiple TXT records; when resolving the challenge's domain, it iterates over the values of the returned TXT records looking for `D-XoblaY4uYM-PO8FrRcbRj45cQF-aYluYM4Xob`.

Unfortunately, the function call

```go
GetTXTForDomain("da81d9b9-4ed1-5343-9a44-4e78f9ead5d9")
```

will only return the two first TXT values instead of the expected three:

```sh
D-XoblaY4uYM-PO8FrRcbRj45cQF-aYluYM4Xob     # ✅ _acme-challenge.mael.pw.
D-52Wm4V7xoUpGax-F8FrPO45cQRcbRj-XoblaY4uYM # ✅ _acme-challenge.blog.mael.pw.
                                            # ❌ _acme-challenge.minio.mael.pw.
```

This means the challenge for `_acme-challenge.minio.mael.pw.` will never succeed.

As detailed in [joohoi/acme-dns#76](https://github.com/joohoi/acme-dns/issues/76), the reason but only returning two TXT records is due to limitations of some ACME servers (Boulder, which is Let's Encrypt's ACME backend, can [only](https://github.com/joohoi/acme-dns/issues/76#issuecomment-385235833) process DNS responses 4096 bytes long).

I will update the page to say that at most two subdomains should ever be associated with one set of credentials.

Reply:
Done in c31897cc2b199ad4a956e178482c8e119a1b1f3a. I also added a comment to the initial issue https://github.com/jetstack/cert-manager/issues/3610#issuecomment-849792721 with the above analysis/example.